server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java

```diff
@@ -0,0 +1,138 @@
+import org.elasticsearch.cluster.metadata.ReservedStateMetadata;
+import org.elasticsearch.reservedstate.ReservedClusterStateHandler;
+import org.elasticsearch.reservedstate.TransformState;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.reservedstate.service.ErrorState.unwrapException;
+
+/**
+ * Generic task to update and reserve parts of the cluster state
+ *
+ * <p>
+ * Reserved cluster state can only be modified by using the {@link ReservedClusterStateController}. Updating
+ * the reserved cluster state through REST APIs is not permitted.
+ */
+public class ReservedStateUpdateTask implements ClusterStateTaskListener {
+    private static final Logger logger = LogManager.getLogger(ReservedStateUpdateTask.class);
+
+    private final String namespace;
+    private final ReservedStateChunk stateChunk;
+    private final Map<String, ReservedClusterStateHandler<?>> handlers;
+    private final Collection<String> orderedHandlers;
+    private final Consumer<ErrorState> errorReporter;
+    private final ActionListener<ActionResponse.Empty> listener;
+
+    public ReservedStateUpdateTask(
+        String namespace,
+        ReservedStateChunk stateChunk,
+        Map<String, ReservedClusterStateHandler<?>> handlers,
+        Collection<String> orderedHandlers,
+        Consumer<ErrorState> errorReporter,
+        ActionListener<ActionResponse.Empty> listener
+    ) {
+        this.namespace = namespace;
+        this.stateChunk = stateChunk;
+        this.handlers = handlers;
+        this.orderedHandlers = orderedHandlers;
+        this.errorReporter = errorReporter;
+        this.listener = listener;
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+        listener.onFailure(e);
+    }
+
+    ActionListener<ActionResponse.Empty> listener() {
+        return listener;
+    }
+
+    protected ClusterState execute(ClusterState state) {
+        ReservedStateMetadata existingMetadata = state.metadata().reservedStateMetadata().get(namespace);
+        Map<String, Object> reservedState = stateChunk.state();
+        ReservedStateVersion reservedStateVersion = stateChunk.metadata();
+
+        var reservedMetadataBuilder = new ReservedStateMetadata.Builder(namespace).version(reservedStateVersion.version());
+        List<String> errors = new ArrayList<>();
+
+        for (var handlerName : orderedHandlers) {
+            ReservedClusterStateHandler<?> handler = handlers.get(handlerName);
+            try {
+                Set<String> existingKeys = keysForHandler(existingMetadata, handlerName);
+                TransformState transformState = handler.transform(reservedState.get(handlerName), new TransformState(state, existingKeys));
+                state = transformState.state();
```

Comment:
Each handler is create it's own ClusterState. Can we do this more efficiently? Building cluster state is not cheap. Could we instead create a ClusterState.Builder here (or perhaps just the customs and metadata builders), and that is passed to the transform method? I think the TransformState would not be needed, the method could just return the new set of keys.

Reply:
I can't fully use ClusterState.Builder without more significant refactor of how the TransportActions we are wrapping for this work. All of the transport actions we use for the reserved state handlers take cluster state in and return a new state back. However, I really like the idea of using the builder instead of the TransportState. This way we can remove that TransportState class, we can start using the builder and then we can slowly improve the transport actions to maybe work with a builder in the future, to improve efficiency.

Reply:
I think I said that too quickly, I think using the builder at the moment without refactoring all of the transport actions will be more overhead. I'll rip out transport state though, I can simply use Tuple.

Reply:
I'm ok with waiting on this optimization, but I do think it is something we should look into. Creating cluster state is relatively expensive.

Reply:
I'll follow-up with a PR to optimize this after I check that it's OK to refactor the Transport node actions to use a builder.